File: algos/solutions/a2fastslowpointers/PY/findduplicate/algo.py
# algo.py

def find_duplicate(nums):
    """
    Finds the duplicate number in the given array using Floyd's cycle detection algorithm.
    """
    slow, fast = nums[0], nums[0]

    # Phase 1: Finding the intersection point
    while True:
        slow = nums[slow]
        fast = nums[nums[fast]]
        if slow == fast:
            break

    # Phase 2: Finding the entrance to the cycle
    slow = nums[0]
    while slow != fast:
        slow = nums[slow]
        fast = nums[fast]

    return slow

File: algos/solutions/a2fastslowpointers/JS/findduplicate/algo.js
// findduplicate/algo.js

function findDuplicate(nums) {
    let fast = nums[0];
    let slow = nums[0];

    while (true) {
        slow = nums[slow];
        fast = nums[nums[fast]];

        if (slow === fast) {
            break;
        }
    }

    slow = nums[0];

    while (slow !== fast) {
        slow = nums[slow];
        fast = nums[fast];
    }

    return fast;
}

module.exports = { findDuplicate };
